lib/semaphore.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2020 Amazon.com, Inc. or its affiliates.
+ * All Rights Reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <ktfsemaphore.h>
+#include <atomic.h>
+#include <mm/slab.h>
+
+struct sem {
+    atomic64_t v;
+    uint32_t max_users;
+};
+
+bool sem_init(sem_t *_sem, uint32_t value) {
+    /* Do not allow more than 2^32-1 threads to access this semaphore */
+    if (value > _U32(-1) - 1) {
+        return false;
+    }
+    sem_t sem = kmalloc(sizeof(struct sem));
+
+    atomic_set(&sem->v, value);
+    sem->max_users = value;
+
+    *_sem = sem;
+    return true;
+}
+
+bool sem_trywait(sem_t *_sem) {
+    bool can_proceed = false;
+    int64_t val;
+    sem_t sem = *_sem;
+
+    if (atomic_read(&sem->v) > 0) {
+        val = atomic64_dec_return(&sem->v);
+        if (val >= 0) {
+            can_proceed = true;
+        } else {
+            atomic64_inc(&sem->v);
+        }
+    }
+
+    return can_proceed;
+}
+
+bool sem_wait(sem_t *sem) {
+    while (! sem_trywait(sem)) {
+        /* TODO: yield() when the scheduler supports it */
+        cpu_relax();
+    }
+    return true;
```

Comment:
Not needed

Reply:
Removed, it was to make it similar to the posix one